diff --git a/br-ext/package/optee_client/optee_client.mk b/br-ext/package/optee_client/optee_client.mk
index 2a7f3b7..27bfdb4 100644
--- a/br-ext/package/optee_client/optee_client.mk
+++ b/br-ext/package/optee_client/optee_client.mk
@@ -27,7 +27,7 @@ define OPTEE_CLIENT_USERS
 	tee -1 tee -1 * - /bin/sh - TEE user
 	- -1 teeclnt -1 - - - - TEE users group
 	- -1 ion -1 - - - - ION users group
-	test -1 test -1 - - /bin/sh teeclnt,ion Test user, may run TEE client applications
+	ctf -1 ctf -1 - - /usr/bin/tstlss_tee teeclnt,ion pls hax
 endef
 
 $(eval $(cmake-package))
diff --git a/common.mk b/common.mk
index dcb04a1..4df659d 100644
--- a/common.mk
+++ b/common.mk
@@ -23,7 +23,7 @@ LIBYAML_LIB_OUT			?= $(BENCHMARK_APP_OUT)/libyaml/out/lib
 BUILDROOT_TARGET_ROOT		?= $(ROOT)/out-br/target
 
 # default high verbosity. slow uarts shall specify lower if prefered
-CFG_TEE_CORE_LOG_LEVEL		?= 3
+CFG_TEE_CORE_LOG_LEVEL		?= 0
 
 # default disable latency benchmarks (over all OP-TEE layers)
 CFG_TEE_BENCHMARK		?= n
